### PR TITLE
Restore handling of null response

### PR DESCRIPTION
**Pull request by natebosch**
Opt-in code cannot pass a null response, however opt-out code can.
Retain check for a null `Response`, cast to `dynamic` to silence
analyzer warning.

Add a test that specifically opts out of null safety to test this
behavior.

### CODE REVIEW

**Comment by natebosch** [comment]
This does copy/paste a fair bit of test utility code in that file. I wasn't sure if it was worth it to pull it out into the test utils - let me know if you think I should.

File: test/shelf_io_test.dart [hunk 1]
```diff
@@ -42,21 +42,6 @@ void main() {
...
 
-  test('sync null response leads to a 500', () async {
-    await _scheduleServer((request) => Response.internalServerError());
-
-    var response = await _get();
-    expect(response.statusCode, HttpStatus.internalServerError);
-    expect(response.body, 'Internal Server Error');
-  });
-
-  test('async null response leads to a 500', () async {
```
Comment (kevmoo) [inline test/shelf_io_test.dart:53]: isn't this a valid test in sound mode?

**Review by kevmoo — APPROVED** [review]

File: test/shelf_io_test.dart (hunk 1 above)
Comment (natebosch) [reply test/shelf_io_test.dart:53]: It passes for the wrong reason.

What this test demonstrates in sound mode is that the `Future.value` constructor has a bad API for null safety, and the `Future.value(null)` is acting like a `Future.error` in sound mode. We already have that behavior covered in sound mode by the test `'async error leads to a 500'`.

Comment (kevmoo) [reply test/shelf_io_test.dart:53]: 👍

Comment (natebosch) [reply test/shelf_io_test.dart:53]: I'll use `async => null` instead of `Future.value(null)` to make this more clear why it's only an opt-out test. The former would have a static error in opt-in code.

**Review by jakemac53 — APPROVED** [review]
